sqlite/vdbeaux.c

```diff
@@ -5292,6 +5292,47 @@ Mem* sqlite3GetCachedResultRow(sqlite3_stmt *pStmt, int *nColumns)
 
   return NULL;
 }
+
```

Comment:
When using the internal memory allocation routines (e.g. sqlite3Malloc, et al), the database mutex should be held; otherwise, an assert() will be fired (because the usage is not thread safe.

Reply:
why do we need a mutex for single-threaded execution?

Reply:
When the per-thread malloc is disabled, the SQLite memory allocator is used and its mutex's must be respected.

Reply:
this is done during sqlite normal processing, so the mutex is acquired if enabled.